app/models/user.rb

```diff
@@ -1,2 +1,17 @@
 class User < ApplicationRecord
+  VALID_EMAIL_REGEX = Settings.models.user.email_regex
+
+  before_save :downcase
+  validraates :name, presence: true, length: {maximum: Settings.models.user.namevalidates}
+  validates :email, presence: true, length: {maximum: Settings.models.user.emaillvalidates},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: {case_sensitive: false}
+  validates :password, presence: true, length: {minimum:  Settings.models.user.passvalidates}
+
+  has_secure_password
```

Comment:
Method hes_serure_password này ntn ?

Reply:
e đọc trong sách thì nó  là để mã hóa mật khẩu a ạ

Reply:
@phamhuuluan chi tiết hơn đi em, thế này sơ sài quá

Reply:
là giải thích chi tiết hơn ạ

Reply:
+, Tải bcrypt-rubyGem và yêu cầu bcrypt. bcrypt là một chức năng băm an toàn.
+, Thêm một thuộc tính chỉ đọc vào mô hình được gọi password.
+, Xác nhận rằng mật khẩu được xác nhận bởi một trường khác, được gọi 
      password_confirmation. Nói cách khác, bạn phải nhập mật khẩu hai lần để xác nhận nó.
+, Đảm bảo rằng password_digest có mặt trước khi mô hình được lưu.
+,Tải các phương thức cá thể , trong trường hợp này là authenticate(trả về true nếu mật khẩu đúng, ngược lại là sai) và password=mã hóa mật khẩu đã truyền vào thuộc tính password_digest.
+, Nếu phương thức có các thuộc tính được bảo vệ theo mặc định, điều này cũng sẽ thêm password_digestvào danh sách các thuộc tính được bảo vệ đó. (Do đó ngăn không cho nó được chỉ định hàng loạt.)